File: game-frontend/src/Command/putBlock.ts

import {CommandBase, Undoable} from 'interacto';
import {Tile, Block, Type} from '../Model/tile';
import {Game} from '../Model/game';

export class PutBlock extends CommandBase implements Undoable  {
  private block: Block;
  private  tile: Tile;
  private oldScore: number;
  private newScore: number;
  private game: Game;
  private nbUpdates: number;



  constructor(game: Game, block: Block, tile: Tile) {
    super();
    this.game = game;
    this.block = block;
    this.oldScore = this.game.getScore();
    this.newScore = this.oldScore + this.game.updateScore(tile, block);
    this.nbUpdates = 0;
    if (!tile.Occupied && tile.Type === Type.GRASS) {
      this.tile = tile;
    } else {
      throw new Error('erreur');
    }
  }


  execution(): void {
      this.tile.block = this.block;
      this.tile.Occupied = true;
      this.game.setScore(this.newScore);
      this.game.getPlayer().inventaire.removeBlock(this.block);
      while (this.game.getScore() > this.game.getScorLimit()) {
        this.game.updateScoreLimit();
        this.nbUpdates++;
      }
}

get  chosenBlock(): Block {
  return this.block;
}

get chosenTile(): Tile {
  return this.tile;
}


 undo(): void{
  this.tile.block = (Block.NOTHING);
  this.tile.Occupied = (false);
  this.game.setScore(this.oldScore);
  this.game.getPlayer().inventaire.addBlock(this.block);
  for (let i = 0; i < this.nbUpdates; i++) {
    this.game.decreaseScoreLimit();
  }
  this.nbUpdates = 0;
}

  get OldScore(): number{
    return this.oldScore;
}

  get NewScore(): number {
    return this.newScore;
}

  redo(): void {
    this.tile.block = this.block;
    this.tile.Occupied = true;
    this.game.setScore(this.newScore);
    this.game.getPlayer().inventaire.removeBlock(this.block);
    while ( this.game.getScore() > this.game.getScorLimit()) {
      this.game.updateScoreLimit();
      this.nbUpdates++;
    }
  }

  getUndoName(): string {
    return 'put block';
  }
}

File: game-backend/src/main/java/game/Commands/PutBlock.java
package game.Commands;

import game.model.Block;
import game.model.Game;
import game.model.Tile;
import game.model.Type;

public class PutBlock implements Command, Undoable {
    private final Block block;
    private final Tile tile;
    private final int oldScore;
    private final int newScore;
    private final Game game;


    /**
     * create a new command to place a Block on a given Tile of the Map
     * @param game the game the player is on
     * @param block the block to place
     * @param tile the tile on which the block will be put on
     */
    public PutBlock(final Game game, final Block block, final Tile tile) {
        this.game = game;
        this.block = block;
        this.oldScore = game.getScore();
        this.newScore = oldScore + game.updateScore(tile, block);
        if(!tile.getOccupied() && tile.getType() == Type.GRASS) {
            this.tile = tile;
        } else {
            throw new IllegalArgumentException();
        }
    }

    /**
     * set the block on the given tile and update the score
     */
    @Override
    public void execute() {
        tile.setBlock(block);
        tile.setOccupied(true);
        game.setScore(newScore);
        game.getPlayer().getInventaire().removeBlock(block);
        game.updateScoreLimit();
        redo();
    }

    public Block getBlock() {
        return block;
    }

    public Tile getTile() {
        return tile;
    }

    @Override
    public void undo() {
        tile.setBlock(Block.NOTHING);
        tile.setOccupied(false);
        game.setScore(oldScore);
        game.getPlayer().getInventaire().addBlock(block);
        game.decreaseScoreLimit();
    }

    public int getOldScore() {
        return oldScore;
    }

    public int getNewScore() {
        return newScore;
    }

    @Override
    public void redo() {
        tile.setBlock(block);
        tile.setOccupied(true);
        game.setScore(newScore);
        game.getPlayer().getInventaire().removeBlock(block);
        game.updateScoreLimit();
    }
}
